loop.c

```diff
@@ -145,34 +157,37 @@ struct loop_timer *loop_add_timer(struct loop *loop, int ms,
 	}
 	timer->expiry.tv_nsec += nsec;
 
-	list_add(loop->timers, timer);
+	wl_list_insert(&loop->timers, &timer->link);
 
 	return timer;
 }
 
 bool loop_remove_fd(struct loop *loop, int fd) {
-	for (int i = 0; i < loop->fd_length; ++i) {
-		if (loop->fds[i].fd == fd) {
-			free(loop->fd_events->items[i]);
-			list_del(loop->fd_events, i);
+	int fd_index = 0;
+	struct loop_fd_event *event = NULL, *tmp_event = NULL;
+	wl_list_for_each_safe(event, tmp_event, &loop->fd_events, link) {
+		if (loop->fds[fd_index].fd == fd) {
+			wl_list_remove(&event->link);
+			free(event);
 
 			loop->fd_length--;
-			memmove(&loop->fds[i], &loop->fds[i + 1],
-					sizeof(struct pollfd) * (loop->fd_length - i));
-
+			memmove(&loop->fds[fd_index], &loop->fds[fd_index + 1],
+					sizeof(struct pollfd) * (loop->fd_length - fd_index));
 			return true;
 		}
-	}
+		++fd_index;
+	};
 	return false;
 }
 
-bool loop_remove_timer(struct loop *loop, struct loop_timer *timer) {
-	for (int i = 0; i < loop->timers->length; ++i) {
-		if (loop->timers->items[i] == timer) {
-			list_del(loop->timers, i);
+bool loop_remove_timer(struct loop *loop, struct loop_timer *remove) {
+	struct loop_timer *timer = NULL, *tmp_timer = NULL;
+	wl_list_for_each_safe(timer, tmp_timer, &loop->timers, link) {
+		if (timer == remove) {
+			wl_list_remove(&timer->link);
 			free(timer);
 			return true;
 		}
-	}
+	};
```

Comment:
Style: extra `;` here

Reply:
Fixed... I'm pretty sure gcc was giving me errors about missing semicolons and that's why I added them, but appears to be working now so whatever...